File: python/samples/sensor_data_flow.py
"""Sensor Data Flow Example: Proprietary -> ISO19848.

This example demonstrates transforming proprietary sensor data into an
ISO19848-compliant TimeSeriesDataPackage.

Note: This is a simplified example for illustration purposes and may not
cover all aspects of a production ISO19848 implementation.

The DataChannelList serves as the mapping between system IDs (short_id)
and ISO19848 LocalIds - no separate mapping configuration is needed.

Flow:
    1. Load ISO19848 DataChannelList (defines valid channels + mappings)
    2. Receive proprietary sensor readings with system IDs
    3. Filter validates readings against DataChannelList
    4. Group channels by update_cycle into TabularData (ISO19848 recommendation)
    5. Produce ISO19848 TimeSeriesDataPackage
    6. Serialize to standard JSON format
"""

from collections import defaultdict
from dataclasses import dataclass, field
from datetime import datetime, timezone

from samples.iso19848_json import create_sample_data_channel_list

from vista_sdk.imo_number import ImoNumber
from vista_sdk.system_text_json import JsonExtensions, Serializer
from vista_sdk.transport.data_channel.data_channel import (
    DataChannel,
    DataChannelListPackage,
)
from vista_sdk.transport.ship_id import ShipId
from vista_sdk.transport.time_series_data import (
    DataChannelId,
    TabularData,
    TimeSeriesData,
    TimeSeriesDataPackage,
)
from vista_sdk.transport.time_series_data import Header as TimeSeriesHeader
from vista_sdk.transport.time_series_data import Package as TimeSeriesPackage
from vista_sdk.transport.time_series_data.time_series_data import (
    TabularDataSet,
    TimeSpan,
)


@dataclass
class SensorReading:
    """A proprietary sensor reading with system ID."""

    system_id: str  # The short_id used internally
    value: float
    timestamp: datetime
    quality: str = "Good"


@dataclass
class ISO19848Filter:
    """Filter that transforms proprietary sensor data to ISO19848 format.

    Uses the DataChannelList as the source of truth for:
    - Valid channel IDs (short_id)
    - Mapping from short_id to LocalId
    - Update cycle for grouping channels (ISO19848 optimization)

    Any reading with an unknown system_id is rejected.
    Channels with the same update_cycle are grouped into one TabularData.
    """

    data_channel_list_package: DataChannelListPackage
    _buffer: dict[str, list[SensorReading]] = field(default_factory=dict)

    def _get_update_cycle(self, dc: DataChannel) -> float | None:
        """Get the update cycle for a data channel."""
        return dc.property_.data_channel_type.update_cycle

    def receive(self, reading: SensorReading) -> bool:
        """Receive a sensor reading. Returns True if accepted."""
        result, _ = (
            self.data_channel_list_package.data_channel_list.try_get_by_short_id(
                reading.system_id
            )
        )
        if not result:
            print(f"    ✗ Rejected: {reading.system_id} (unknown channel)")
            return False

        if reading.system_id not in self._buffer:
            self._buffer[reading.system_id] = []

        self._buffer[reading.system_id].append(reading)
        print(f"    ✓ Accepted: {reading.system_id} = {reading.value}")
        return True

    def flush(self, ship_id: ShipId) -> TimeSeriesDataPackage | None:
        """Flush buffered readings to a TimeSeriesDataPackage.

        Groups channels by update_cycle into TabularData blocks.
        This is recommended by ISO19848 to reduce payload size.
        """
        if not self._buffer:
            return None

        # Reference the specific DataChannelList version (recommended)
        configuration = self.data_channel_list_package.package.header.data_channel_list_id.as_time_series_reference()  # noqa: E501

        # Group channels by update_cycle
        channels_by_cycle: dict[float | None, list[tuple[DataChannel, str]]] = (
            defaultdict(list)
        )

        for system_id in self._buffer:
            result, dc = (
                self.data_channel_list_package.data_channel_list.try_get_by_short_id(
                    system_id
                )
            )
            if result and dc is not None:
                update_cycle = self._get_update_cycle(dc)
                channels_by_cycle[update_cycle].append((dc, system_id))

        # Build one TabularData per update_cycle group
        tabular_data_list: list[TabularData] = []

        for update_cycle, channel_pairs in channels_by_cycle.items():
            # Collect all unique timestamps across channels in this group
            all_timestamps: set[datetime] = set()
            for _, system_id in channel_pairs:
                for buf_reading in self._buffer[system_id]:
                    all_timestamps.add(buf_reading.timestamp)

            sorted_timestamps = sorted(all_timestamps)

            # Build data channel IDs list for this group
            data_channel_ids = [
                DataChannelId.from_local_id(dc.data_channel_id.local_id)
                for dc, _ in channel_pairs
            ]

            # Build data sets - one per timestamp, with values for all channels
            data_sets: list[TabularDataSet] = []
            for ts in sorted_timestamps:
                values: list[str] = []
                qualities: list[str] = []

                for _, system_id in channel_pairs:
                    # Find reading for this channel at this timestamp
                    reading = next(
                        (r for r in self._buffer[system_id] if r.timestamp == ts),
                        None,
                    )
                    if reading:
                        values.append(str(reading.value))
                        qualities.append(reading.quality)
                    else:
                        # No reading for this channel at this timestamp
                        values.append("")
                        qualities.append("Bad")

                data_sets.append(
                    TabularDataSet(time_stamp=ts, value=values, quality=qualities)
                )

            cycle_str = f"{update_cycle}s" if update_cycle else "None"
            print(
                f"    TabularData: {len(channel_pairs)} channels, "
                f"{len(data_sets)} rows (update_cycle={cycle_str})"
            )

            tabular_data_list.append(
                TabularData(data_channel_ids=data_channel_ids, data_sets=data_sets)
            )

        # Build the package
        all_timestamps_list = [
            r.timestamp for readings in self._buffer.values() for r in readings
        ]

        package = TimeSeriesPackage(
            header=TimeSeriesHeader(
                ship_id=ship_id,
                time_span=TimeSpan(
                    start=min(all_timestamps_list),
                    end=max(all_timestamps_list),
                ),
                author="ISO19848Filter",
            ),
            time_series_data=[
                TimeSeriesData(
                    data_configuration=configuration,
                    tabular_data=tabular_data_list,
                    event_data=None,
                )
            ],
        )

        self._buffer.clear()
        return TimeSeriesDataPackage(package)


def main() -> None:
    """Demonstrate sensor data flowing through an ISO19848 filter."""
    print("=" * 60)
    print("Sensor Data Flow: Proprietary -> ISO19848")
    print("=" * 60)

    # Step 1: Load DataChannelList (defines valid channels + mappings)
    print("\n[1] Loading DataChannelList...")
    dcl = create_sample_data_channel_list()
    print(f"    Loaded {len(dcl.data_channel_list)} channels:")
    for dc in dcl.data_channel_list:
        cycle = dc.property_.data_channel_type.update_cycle
        cycle_str = f"{cycle}s" if cycle else "None"
        print(f"      {dc.data_channel_id.short_id} (update_cycle={cycle_str})")

    # Step 2: Create the filter
    print("\n[2] Creating ISO19848 filter...")
    iso_filter = ISO19848Filter(dcl)
    print("    Filter ready")
    print("    (Channels with same update_cycle will be grouped)")

    # Step 3: Receive sensor readings
    print("\n[3] Receiving sensor readings...")
    base_time = datetime(2024, 6, 15, 10, 0, 0, tzinfo=timezone.utc)

    # Simulate readings from multiple sensors at same timestamps
    readings = [
        # Temperature readings (1.0s update cycle)
        SensorReading("TEMP001", 45.2, base_time),
        SensorReading("TEMP001", 46.1, base_time.replace(minute=5)),
        # Power readings (same 1.0s update cycle - will be grouped!)
        SensorReading("PWR001", 2500.0, base_time),
        SensorReading("PWR001", 2650.5, base_time.replace(minute=5)),
        # Alert channel (different update cycle)
        SensorReading("ALT001", 1.0, base_time),
        # Unknown sensor (will be rejected)
        SensorReading("UNKNOWN", 123.4, base_time),
    ]

    for reading in readings:
        iso_filter.receive(reading)

    # Step 4: Flush to TimeSeriesDataPackage
    print("\n[4] Generating TimeSeriesDataPackage (grouped by update_cycle)...")
    ship_id = ShipId(ImoNumber(1234567))
    package = iso_filter.flush(ship_id)

    if package:
        print("    ✓ Package created")
        if package.package.header:
            h = package.package.header
            print(f"    Ship: {h.ship_id}")
            if h.time_span:
                print(f"    TimeSpan: {h.time_span.start} - {h.time_span.end}")

    # Step 5: Serialize to JSON
    print("\n[5] Serializing to JSON...")
    if package:
        json_dto = JsonExtensions.TimeSeriesData.to_json_dto(package)
        json_str = Serializer.serialize(json_dto)
        byte_count = len(json_str.encode("utf-8"))
        print(f"    ✓ {byte_count} bytes")
        print(f"    Preview: {json_str[:80]}...")

    print("\n" + "=" * 60)
    print("Done! Data is now ISO19848-compliant.")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: js/samples/SensorDataFlow.ts
/*
    Sensor Data Flow Example: Proprietary -> ISO19848

    This example demonstrates transforming proprietary sensor data into an
    ISO19848-compliant TimeSeriesDataPackage.

    Note: This is a simplified example for illustration purposes and may not
    cover all aspects of a production ISO19848 implementation.

    The DataChannelList serves as the mapping between system IDs (short_id)
    and ISO19848 LocalIds - no separate mapping configuration is needed.

    Flow:
        1. Load ISO19848 DataChannelList (defines valid channels + mappings)
        2. Receive proprietary sensor readings with system IDs
        3. Filter validates readings against DataChannelList
        4. Group channels by update_cycle into TabularData (ISO19848 recommendation)
        5. Produce ISO19848 TimeSeriesDataPackage
        6. Serialize to standard JSON format
*/

import {
    DataChannelId,
    DataChannelList,
    JSONExtensions,
    JSONSerializer,
    LocalId,
    ShipId,
    TimeSeries,
    Version,
} from "dnv-vista-sdk";

// =========================================================================
// Types
// =========================================================================

interface SensorReading {
    systemId: string;
    value: number;
    timestamp: Date;
    quality: string;
}

// =========================================================================
// ISO19848 Filter
// =========================================================================

class ISO19848Filter {
    private readonly dcl: DataChannelList.DataChannelListPackage;
    private readonly shortIdMap: Map<string, DataChannelList.DataChannel>;
    private readonly buffer: Map<string, SensorReading[]> = new Map();

    constructor(dcl: DataChannelList.DataChannelListPackage) {
        this.dcl = dcl;
        // Build a shortId lookup map
        this.shortIdMap = new Map();
        for (const dc of dcl.package.dataChannelList.dataChannel) {
            if (dc.dataChannelId.shortId) {
                this.shortIdMap.set(dc.dataChannelId.shortId, dc);
            }
        }
    }

    receive(reading: SensorReading): boolean {
        if (!this.shortIdMap.has(reading.systemId)) {
            console.log(
                `    ✗ Rejected: ${reading.systemId} (unknown channel)`,
            );
            return false;
        }

        if (!this.buffer.has(reading.systemId)) {
            this.buffer.set(reading.systemId, []);
        }
        this.buffer.get(reading.systemId)!.push(reading);
        console.log(`    ✓ Accepted: ${reading.systemId} = ${reading.value}`);
        return true;
    }

    async flush(
        shipId: ShipId,
    ): Promise<TimeSeries.TimeSeriesDataPackage | undefined> {
        if (this.buffer.size === 0) return undefined;

        // Group channels by update_cycle
        const channelsByCycle = new Map<
            string,
            { dc: DataChannelList.DataChannel; systemId: string }[]
        >();

        for (const systemId of this.buffer.keys()) {
            const dc = this.shortIdMap.get(systemId);
            if (!dc) continue;

            const updateCycle = dc.property.dataChannelType.updateCycle;
            const cycleKey =
                updateCycle !== undefined ? updateCycle.toString() : "none";

            if (!channelsByCycle.has(cycleKey)) {
                channelsByCycle.set(cycleKey, []);
            }
            channelsByCycle.get(cycleKey)!.push({ dc, systemId });
        }

        // Build one TabularData per update_cycle group
        const tabularDataList: TimeSeries.TabularData[] = [];

        for (const [updateCycle, channelPairs] of channelsByCycle) {
            // Collect all unique timestamps
            const allTimestamps = new Set<number>();
            for (const { systemId } of channelPairs) {
                for (const reading of this.buffer.get(systemId)!) {
                    allTimestamps.add(reading.timestamp.getTime());
                }
            }
            const sortedTimestamps = [...allTimestamps].sort();

            // Build data channel IDs using parseAsync
            const dataChannelIds: DataChannelId[] = [];
            for (const { dc } of channelPairs) {
                dataChannelIds.push(
                    await DataChannelId.parseAsync(
                        dc.dataChannelId.localId.toString(),
                    ),
                );
            }

            // Build data sets
            const dataSets: TimeSeries.TabularDataSet[] = [];
            for (const ts of sortedTimestamps) {
                const values: string[] = [];
                const qualities: string[] = [];

                for (const { systemId } of channelPairs) {
                    const reading = this.buffer
                        .get(systemId)!
                        .find((r) => r.timestamp.getTime() === ts);
                    if (reading) {
                        values.push(reading.value.toString());
                        qualities.push(reading.quality);
                    } else {
                        values.push("");
                        qualities.push("Bad");
                    }
                }

                dataSets.push({
                    timeStamp: new Date(ts),
                    value: values,
                    quality: qualities,
                });
            }

            const cycleStr =
                updateCycle === "none" ? "None" : `${updateCycle}s`;
            console.log(
                `    TabularData: ${channelPairs.length} channels, ${dataSets.length} rows (update_cycle=${cycleStr})`,
            );

            tabularDataList.push({
                dataChannelId: dataChannelIds,
                dataSet: dataSets,
            });
        }

        // Build the package
        const allTimestampsList: Date[] = [];
        for (const readings of this.buffer.values()) {
            for (const r of readings) {
                allTimestampsList.push(r.timestamp);
            }
        }

        const minTime = new Date(
            Math.min(...allTimestampsList.map((d) => d.getTime())),
        );
        const maxTime = new Date(
            Math.max(...allTimestampsList.map((d) => d.getTime())),
        );

        const pkg: TimeSeries.TimeSeriesDataPackage = {
            package: {
                header: {
                    shipId,
                    timeSpan: { start: minTime, end: maxTime },
                    author: "ISO19848Filter",
                },
                timeSeriesData: [
                    {
                        dataConfiguration: {
                            id: this.dcl.package.header.dataChannelListId.id,
                            timeStamp:
                                this.dcl.package.header.dataChannelListId
                                    .timestamp,
                        },
                        tabularData: tabularDataList,
                    },
                ],
            },
        };

        this.buffer.clear();
        return pkg;
    }
}

// =========================================================================
// Helper: Create sample DataChannelList
// =========================================================================
async function createSampleDataChannelList(): Promise<DataChannelList.DataChannelListPackage> {
    const localId1 = await LocalId.parseAsync(
        "/dnv-v2/vis-3-4a/411.1-1/C101.63/S206/meta/qty-temperature/cnt-cooling.air",
    );
    const localId2 = await LocalId.parseAsync(
        "/dnv-v2/vis-3-4a/511.15-1/E32/meta/qty-power",
    );
    const localId3 = await LocalId.parseAsync(
        "/dnv-v2/vis-3-4a/621.22i/S110/meta/cnt-diesel.oil/cmd-stop",
    );

    const shipId = ShipId.parse("IMO1234567");
    const timestamp = new Date("2024-01-01T00:00:00Z");

    return {
        package: {
            header: {
                shipId,
                dataChannelListId: {
                    id: "DataChannelList-v1",
                    version: Version.parse("1.0"),
                    timestamp,
                },
                versionInformation: {
                    namingRule: "dnv",
                    namingSchemeVersion: "v2",
                    referenceUrl: "https://docs.vista.dnv.com",
                },
                author: "Vista SDK Sample",
                dateCreated: timestamp,
            },
            dataChannelList: {
                dataChannel: [
                    {
                        dataChannelId: {
                            localId: localId1,
                            shortId: "TEMP001",
                        },
                        property: {
                            dataChannelType: {
                                type: "Inst",
                                updateCycle: 1.0,
                            },
                            format: {
                                type: "Decimal",
                                restriction: {
                                    fractionDigits: 1,
                                    maxInclusive: 200.0,
                                    minInclusive: -50.0,
                                },
                            },
                            range: { low: 0.0, high: 150.0 },
                            unit: {
                                unitSymbol: "°C",
                                quantityName: "Temperature",
                            },
                            qualityCoding: "OPC_QUALITY",
                            name: "Main Engine Air Cooler Temperature",
                        },
                    },
                    {
                        dataChannelId: {
                            localId: localId2,
                            shortId: "PWR001",
                        },
                        property: {
                            dataChannelType: {
                                type: "Inst",
                                updateCycle: 1.0,
                            },
                            format: { type: "Decimal" },
                            range: { low: 0.0, high: 10000.0 },
                            unit: {
                                unitSymbol: "kW",
                                quantityName: "Power",
                            },
                            name: "Generator Power Output",
                        },
                    },
                    {
                        dataChannelId: {
                            localId: localId3,
                            shortId: "ALT001",
                        },
                        property: {
                            dataChannelType: { type: "Alert" },
                            format: {
                                type: "String",
                                restriction: { maxLength: 100 },
                            },
                            alertPriority: "Warning",
                            name: "Diesel Oil Low Level Alert",
                        },
                    },
                ],
            },
        },
    };
}

// =========================================================================
// Main
// =========================================================================
async function main() {
    console.log("=".repeat(60));
    console.log("Sensor Data Flow: Proprietary -> ISO19848");
    console.log("=".repeat(60));

    // Step 1: Load DataChannelList
    console.log("\n[1] Loading DataChannelList...");
    const dcl = await createSampleDataChannelList();
    console.log(
        `    Loaded ${dcl.package.dataChannelList.dataChannel.length} channels:`,
    );
    for (const dc of dcl.package.dataChannelList.dataChannel) {
        const cycle = dc.property.dataChannelType.updateCycle;
        const cycleStr = cycle !== undefined ? `${cycle}s` : "None";
        console.log(
            `      ${dc.dataChannelId.shortId} (update_cycle=${cycleStr})`,
        );
    }

    // Step 2: Create the filter
    console.log("\n[2] Creating ISO19848 filter...");
    const isoFilter = new ISO19848Filter(dcl);
    console.log("    Filter ready");
    console.log("    (Channels with same update_cycle will be grouped)");

    // Step 3: Receive sensor readings
    console.log("\n[3] Receiving sensor readings...");
    const baseTime = new Date("2024-06-15T10:00:00Z");
    const fiveMin = 5 * 60 * 1000;

    const readings: SensorReading[] = [
        // Temperature readings (1.0s update cycle)
        {
            systemId: "TEMP001",
            value: 45.2,
            timestamp: baseTime,
            quality: "Good",
        },
        {
            systemId: "TEMP001",
            value: 46.1,
            timestamp: new Date(baseTime.getTime() + fiveMin),
            quality: "Good",
        },
        // Power readings (same 1.0s update cycle - will be grouped!)
        {
            systemId: "PWR001",
            value: 2500.0,
            timestamp: baseTime,
            quality: "Good",
        },
        {
            systemId: "PWR001",
            value: 2650.5,
            timestamp: new Date(baseTime.getTime() + fiveMin),
            quality: "Good",
        },
        // Alert channel (different update cycle)
        {
            systemId: "ALT001",
            value: 1.0,
            timestamp: baseTime,
            quality: "Good",
        },
        // Unknown sensor (will be rejected)
        {
            systemId: "UNKNOWN",
            value: 123.4,
            timestamp: baseTime,
            quality: "Good",
        },
    ];

    for (const reading of readings) {
        isoFilter.receive(reading);
    }

    // Step 4: Flush to TimeSeriesDataPackage
    console.log(
        "\n[4] Generating TimeSeriesDataPackage (grouped by update_cycle)...",
    );
    const shipId = ShipId.parse("IMO1234567");
    const pkg = await isoFilter.flush(shipId);

    if (pkg) {
        console.log("    ✓ Package created");
        if (pkg.package.header) {
            const h = pkg.package.header;
            console.log(`    Ship: ${h.shipId}`);
            if (h.timeSpan) {
                console.log(
                    `    TimeSpan: ${h.timeSpan.start.toISOString()} - ${h.timeSpan.end.toISOString()}`,
                );
            }
        }
    }

    // Step 5: Serialize to JSON
    console.log("\n[5] Serializing to JSON...");
    if (pkg) {
        const jsonDto = JSONExtensions.TimeSeries.toJsonDto(pkg);
        const jsonStr = JSONSerializer.serializeTimeSeriesData(jsonDto);
        const byteCount = Buffer.byteLength(jsonStr, "utf-8");
        console.log(`    ✓ ${byteCount} bytes`);
        console.log(
            `    Preview: ${jsonStr.substring(0, Math.min(80, jsonStr.length))}...`,
        );
    }

    console.log("\n" + "=".repeat(60));
    console.log("Done! Data is now ISO19848-compliant.");
    console.log("=".repeat(60));
}

main().catch(console.error);
